### PR TITLE
Add support for checkbox questions

### LINKED ISSUE #437: Add Checkbox as another question type

### LINKED ISSUE #547: App.vue throws a null error for user who is not part of any organisation

### LINKED ISSUE #561: 0 was not being included in the OTP value when  the OTP began with 0
Screenshot 2021-12-13 at 11 57 45 AM
in the image above, a digit OTP has been entered but the invalid message is still being shown as the first 0 is not being considered

### LINKED ISSUE #563: prevent input number value from changing upon scrolling
to reproduce, go to app.plio.in and just scroll on the field where you are supposed to enter your mobile number.

https://user-images.githubusercontent.com/13185699/145763464-f3ebce39-6fc7-4da4-b292-253cd5cd4d6a.mov

### PR DESCRIPTION
**Pull request by dalmia**
Fixes #437 
Fixes #547
Fixes #561 
Fixes #563 

## Summary
- added checkbox as another question type to choose from in the ItemEditor
- added checkbox to the question type picker dropdown
- there is a constant conversion from Array to Set for the answer and correct_answer to checkboxes. this is because Set is more efficient as a lookup mechanism and so, it makes it faster to check for existing options and deleting them. However, `Set` is not a valid entry type for the database. It converts `{1,2}` (a `Set` value) to `{}` when sent to the database.
- radio buttons for the options in the item modal is converted to checkboxes for the checkbox question
- edge cases:
    - switching from question type mcq to checkbox: correct answer becomes a list with one element which is the correct answer that was previously selected for the mcq question
    - switching from checkbox to mcq: correct answer becomes the first element in the list of correct answers marked for checkbox
    - if a user tries to unselect all options as the correct answer for checkbox: prevent the last selected correct answer option to be unselected and show a toast that at least one correct answer has to be selected
    - delete correct option: if only one correct answer was selected, mark the first option in the remaining list as the correct answer. otherwise, just remove that option from the list of correct answers.
- restructured some tests to reuse code across tests
- some null checks to cover edge cases not previously covered
- added a `jest.setup.js` file to log detailed error traces during tests which earlier used to look very vague (this is what helped me understand the fixes that I had to add to resolve the errors with tests) 
- updated dockerfile to add node_modules to `/app` as well - this is how I got it to work consistently for me
- using input-text instead of input number for OTP (look at the 3rd issue linked to the PR)
- preventing the scrolling action on the input field for input number component (look at the 4th issue linked to the PR)

## Test Plan

- [x] Test Responsiveness
   - [ ] Laptop (1200px)
   - [ ] Tablet (760px)
   - [ ] Phone (320px)
- [x] Cross-Browser Testing
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Safari (use BrowserStack)
- [x] Local Language Support
- [x] Wrote tests
- [x] Tested locally
- [x] Self-review
- [x] Comments have been added appropriately
- [ ] ~Check for bundle size [here](https://bundlephobia.com/) if adding a package~
- [x] Added relevant details like Labels/Projects/Milestones etc.
- [ ] ~Images have `alt` attributes~
- [ ] ~Any `` tags have `width` and `height` specified~
- [ ] ~Any `target="_blank"` links have `rel="noopener"`~
- [x] Only SVGs are used as images. If PNGs are used, their size has been optimised.
- [ ] ~Any SVGs without text have their `aria-label` attributes set~
- [ ] Tested on staging
- [ ] Tested on an actual physical phone
- [ ] Tested on production

### CODE REVIEW

File: tests/unit/components/Editor/ItemEditor.spec.js [hunk 1]
```diff
@@ -1,347 +1,541 @@
...
+              ? clonedeep(itemDetailList)
+              : newItemDetailList,
+          questionTypeIndex: 2,
+        },
+      });
+    };
+
+    beforeEach(() => {
+      mountWrapper();
     });
```
Comment (deepansh96) [inline tests/unit/components/Editor/ItemEditor.spec.js:224]:
```suggestion
    beforeEach(() => mountWrapper());
```

File: tests/unit/components/Editor/ItemEditor.spec.js [hunk 2]
```diff
@@ -1,347 +1,541 @@
...
+        .setValue("");
+      expect(wrapper.vm.maxCharLimit).toBe(100);
+    });
+
+    it("triggers keyup on max char limit input box", async () => {
+      const maxCharLimitInputKeypress = jest.spyOn(
+        ItemEditor.methods,
+        "maxCharLimitInputKeypress"
+      );
+      mountWrapper();
```
Comment (deepansh96) [inline tests/unit/components/Editor/ItemEditor.spec.js:458]: We're mounting again here?
Shouldn't we unmount the one first which was already mounted?
And why is there a need of doing this?

File: tests/unit/components/Editor/ItemEditor.spec.js [hunk 3]
```diff
@@ -1,347 +1,541 @@
...
-        ],
-        questionTypeIndex: 1,
-      },
+  describe("buttons", () => {
+    it("clicking add item hides item editor", async () => {
+      const removeSelectedItemIndex = jest.spyOn(
+        ItemEditor.methods,
+        "removeSelectedItemIndex"
+      );
+      mountWrapper();
```
Comment (deepansh96) [inline tests/unit/components/Editor/ItemEditor.spec.js:490]: same question as before here - about `mountWrapper()`

File: tests/unit/components/Editor/ItemEditor.spec.js [hunk 4]
```diff
@@ -1,347 +1,541 @@
...
+        type: "mcq",
+        options: ["", ""],
+      });
+
+      it("clicking next item navigates to next item", async () => {
+        const updateSelectedItemIndex = jest.spyOn(
+          ItemEditor.methods,
+          "updateSelectedItemIndex"
+        );
+        mountWrapper(updatedItemList, updatedItemDetailList);
```
Comment (deepansh96) [inline tests/unit/components/Editor/ItemEditor.spec.js:518]: instead of mounting everything again you can just set the props na using `setProps`?

File: tests/unit/components/Items/Question/Body.spec.js [hunk 5]
```diff
@@ -2,266 +2,348 @@ import { mount } from "@vue/test-utils";
...
   });
 
-  it("should render subjective question with default values", () => {
-    const wrapper = mount(Body, {
-      props: {
-        questionType: "subjective",
-      },
+  describe("subjective questions", () => {
+    const mountWrapper = () => {
+      wrapper = mount(Body, {
```
Comment (deepansh96) [inline tests/unit/components/Items/Question/Body.spec.js:59]: So basically you're just mounting a new instance of the component and reassinging it to the wrapper variable. But doesn't this mean that the previously mounted instance will stay mounted?
If that is the case, then it will make tests unnecessarily slow. Best to unmount if we're remounting?

File: tests/unit/components/Items/Question/Body.spec.js [hunk 6]
```diff
@@ -2,266 +2,348 @@ import { mount } from "@vue/test-utils";
...
-        options: options,
-      },
+    it("starts loading image if imageData is passed", () => {
+      const startImageLoading = jest.spyOn(Body.methods, "startImageLoading");
+      mount(Body, {
+        props: {
+          questionType: "mcq",
+          imageData: {
+            url: "mock",
+            alt_text: "mock",
```
Comment (deepansh96) [inline tests/unit/components/Items/Question/Body.spec.js:48]: if we're not using the dummydata files then let's remove them

File: tests/unit/components/Player/ItemModal.spec.js [hunk 7]
```diff
@@ -4,249 +4,389 @@ import ItemModal from "@/components/Player/ItemModal";
...
+    it("should prepare draft responses for each item", () => {
+      expect(wrapper.vm.draftResponses.length).toBe(global.dummyItems.length);
+    });
+
+    it("toggles minimize button", async () => {
+      const toggleMinimizeMock = jest.spyOn(
+        ItemModal.methods,
+        "toggleMinimize"
+      );
+      remount();
```
Comment (deepansh96) [inline tests/unit/components/Player/ItemModal.spec.js:42]: why a remount was needed here?

File: tests/unit/components/Player/ItemModal.spec.js [hunk 8]
```diff
@@ -4,249 +4,389 @@ import ItemModal from "@/components/Player/ItemModal";
...
+        },
       });
+    };
+
+    it("submits question", async () => {
+      let draftResponses = clonedeep(global.dummyItemResponses);
+      const submitQuestion = jest.spyOn(ItemModal.methods, "submitQuestion");
+
+      mountWrapper();
+      // enter some value in the input field
```
Comment (deepansh96) [inline tests/unit/components/Player/ItemModal.spec.js:107]: In this describe, you're not mounting at `beforeEach` but rather mounting at each individual tests, but in the test suite `describe("checkbox questions", () => {`, you have a `beforeEach` and you're mounting in that.
So, any particular reason for NOT beforeEach mounting in `subjective question` describe and mounting beforeEach in `checkbox question` describe?

File: tests/unit/components/Player/ItemModal.spec.js [hunk 9]
```diff
@@ -4,249 +4,389 @@ import ItemModal from "@/components/Player/ItemModal";
...
+
+      // unselect the other option
+      await wrapper.find('[data-test="option-1"]').trigger("click");
+      // the answer should now be invalid
+      expect(wrapper.vm.isAnswerValid).toBeFalsy();
+    });
+
+    it("submits question", async () => {
+      const submitQuestion = jest.spyOn(ItemModal.methods, "submitQuestion");
+      mountWrapper();
```
Comment (deepansh96) [inline tests/unit/components/Player/ItemModal.spec.js:350]: Why are you mounting here again when there's mounting happening in the beforeEach of this test suite?

File: tests/unit/pages/Editor.spec.js [hunk 10]
```diff
@@ -1417,21 +1368,7 @@ describe("Editor.vue", () => {
...
-    const editorDeleteOption = jest.spyOn(Editor.methods, "deleteOption");
-    const deleteSelectedOption = jest.spyOn(
-      Editor.methods,
-      "deleteSelectedOption"
-    );
-    const showCannotDeleteOptionDialog = jest.spyOn(
-      Editor.methods,
-      "showCannotDeleteOptionDialog"
-    );
+  it("deleting options is prevented with only 2 options", async () => {
```
Comment (deepansh96) [inline tests/unit/pages/Editor.spec.js:1371]:
```suggestion
  it("deleting option button is hidden with only 2 options", async () => {
```

File: tests/unit/pages/Editor.spec.js [hunk 11]
```diff
@@ -1522,9 +1401,11 @@ describe("Editor.vue", () => {
         };
       },
     });
+    let updatedDummyItemDetails = clonedeep(global.dummyItemDetails);
+    updatedDummyItemDetails[0].options.push("option 3");
     await wrapper.setData({
       items: clonedeep(global.dummyItems),
-      itemDetails: clonedeep(global.dummyItemDetails),
+      itemDetails: updatedDummyItemDetails,
```
Comment (deepansh96) [inline tests/unit/pages/Editor.spec.js:1408]: why was this needed?

File: tests/unit/pages/Editor.spec.js [hunk 12]
```diff
@@ -1714,14 +1731,14 @@ describe("Editor.vue", () => {
...
-    let createdQuestionResponse = {
+    createdQuestionResponse = {
       data: {
         id: 212,
         item: 212,
         text: "",
         type: "mcq",
         options: ["", ""],
-        correct_answer: "0",
+        correct_answer: 0,
```
Comment (deepansh96) [inline tests/unit/pages/Editor.spec.js:1741]: did we change the data type in case mcq answers from string to integer?

File: tests/unit/pages/embeds/Plio.spec.js [hunk 13]
```diff
@@ -135,197 +145,121 @@ describe("Plio.vue", () => {
...
-    await flushPromises();
-
-    // resolve the `GET` request waiting in the queue (for receiving plio details)
-    // using the fake response data
-    mockAxios.mockResponse(
-      clonedeep(global.dummyPublishedPlio),
-      mockAxios.queue()[0]
-    );
+  describe("session creation", () => {
+    it("session should not be created if unauthenticated when plio request is resolved", async () => {
```
Comment (deepansh96) [inline tests/unit/pages/embeds/Plio.spec.js:149]: in the title it says that session should not be created but in the code it checks if the session has been created?

File: tests/unit/pages/embeds/Plio.spec.js [hunk 14]
```diff
@@ -135,197 +145,121 @@ describe("Plio.vue", () => {
...
-    const wrapper = mount(Plio);
+        await loadPlio();
 
-    // no SSO in the beginning
-    expect(wrapper.vm.isThirdPartyAuth).toBeFalsy();
-    // setting props for SS)
-    await wrapper.setProps({
-      thirdPartyApiKey: "apiKey",
-      thirdPartyUniqueId: "0000000000",
+        // 1 `POST` request should have been made
```
Comment (deepansh96) [inline tests/unit/pages/embeds/Plio.spec.js:195]:
```suggestion
        // 0 `POST` request should have been made
```

File: tests/unit/pages/embeds/Plio.spec.js [hunk 15]
```diff
@@ -135,197 +145,121 @@ describe("Plio.vue", () => {
...
-        return new Promise((resolve) => resolve());
-      });
-    const wrapper = mount(Plio, {
-      props: {
+
+    it("detects SAML SSO", async () => {
+      await mountWrapper();
+      // no SSO in the beginning
+      expect(wrapper.vm.isThirdPartyAuth).toBeFalsy();
+      // setting props for SS)
```
Comment (deepansh96) [inline tests/unit/pages/embeds/Plio.spec.js:209]:
```suggestion
      // setting props for SSO
```

File: tests/unit/pages/embeds/Plio.spec.js [hunk 16]
```diff
@@ -135,197 +145,121 @@ describe("Plio.vue", () => {
...
+
+      mockAxios.mockResponse(
+        { data: global.dummyAccessToken },
+        mockAxios.queue()[0]
+      );
+      await flushPromises();
+      expect(setAccessToken).toHaveBeenCalled();
+      await flushPromises();
+      expect(setActiveWorkspace).toHaveBeenCalled();
+      await flushPromises();
```
Comment (deepansh96) [inline tests/unit/pages/embeds/Plio.spec.js:252]: do we really need 3 flushPromises?

**Review by deepansh96 — CHANGES_REQUESTED** [review]

**Comment by deepansh96** [comment]
Also check this out - https://www.notion.so/avantifellows/Feedback-3d3a811c865f4102af50fdbd17355e96

File: src/components/Editor/ItemEditor.vue [hunk 17]
```diff
@@ -288,6 +298,9 @@ export default {
         iconClass:
           "w-6 h-6 text-primary group-hover:text-white group-disabled:text-primary",
       },
+      // set containing the question types in which options are present
+      questionTypesWithOptions: new Set(["mcq", "checkbox"]),
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/Editor/ItemEditor.vue:302]: convert items wala set thing to array only

File: src/components/Editor/ItemEditor.vue [hunk 18]
```diff
@@ -296,6 +309,13 @@ export default {
       // if the user has not set the limit - reset it back to 100
       if (this.maxCharLimit == "") this.maxCharLimit = 100;
     },
+    isQuestionDropdownShown(value) {
+      const tooltip = document.getElementById("questionTypePickerContainer")._tippy;
+      if (tooltip == undefined) return;
+      if (value) {
+        tooltip.hide();
+      } else tooltip.show();
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/Editor/ItemEditor.vue:317]: use disabled (look at notion doc below for understanding)

File: src/components/Items/Question/Body.vue [hunk 19]
```diff
@@ -113,10 +113,20 @@ export default {
...
+      wrongOptionClass: "text-white bg-red-500",
     };
   },
   watch: {
+    draftAnswer(value) {
+      if (this.submittedAnswer == null) this.subjectiveAnswer = value;
+    },
+    submittedAnswer(value) {
+      this.subjectiveAnswer = value;
+    },
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/Items/Question/Body.vue:129]: check if extra watcher is being called repeatedly when writing answer of subjective - do console.log here + add comment here

File: src/components/Player/ItemModal.vue [hunk 20]
```diff
@@ -225,12 +228,13 @@ export default {
       )
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/Player/ItemModal.vue:228]: isAnswerCorrect

File: src/pages/Editor.vue [hunk 21]
```diff
@@ -4,7 +4,7 @@
     <div
       class="flex flex-col w-full"
       :class="{ 'opacity-30 pointer-events-none': isBackgroundDisabled }"
-      data-test="blurDiv"
+      data-test="container"
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Editor.vue:7]: very generic name

File: src/pages/Embeds/Plio.vue [hunk 22]
```diff
@@ -662,23 +680,29 @@ export default {
      * saves the answer to the question at the current index
      */
     submitQuestion() {
+      let itemResponse = clonedeep(this.itemResponses[this.currentItemIndex]);
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/pages/Embeds/Plio.vue:683]: remove this when converting everything to array

File: tests/unit/components/Player/ItemModal.spec.js (hunk 7 above)
Comment (dalmia) [reply tests/unit/components/Player/ItemModal.spec.js:42]: use with BeforeEach and name it as mountWrapper

File: src/components/Player/ItemModal.vue [hunk 23]
```diff
@@ -298,7 +310,19 @@ export default {
     },
     optionSelected(optionIndex) {
       // invoked when an option is selected
-      this.draftResponses[this.selectedItemIndex] = optionIndex;
+      if (this.isQuestionTypeMCQ) {
+        // for MCQ, simply set the option as the current response
+        this.draftResponses[this.selectedItemIndex] = optionIndex;
```
[the remark this answers is not in the document]
Comment (dalmia) [reply src/components/Player/ItemModal.vue:320]: I'll explain in person why we cannot do this.

File: src/pages/Login.vue [hunk 24]
```diff
@@ -122,14 +125,15 @@
         data-test="phone"
       ></input-number>
       <!-- input box to enter OTP -->
-      <input-number
+      <input-text
```
Comment (deepansh96) [inline src/pages/Login.vue:128]: why

File: src/services/API/RootClient.js [hunk 25]
```diff
@@ -64,6 +65,9 @@ client.interceptors.response.use(
     // logging the error
     console.log(error);
 
+    if (error.config != undefined && error.config.url == otpVerifyEndpoint)
```
Comment (deepansh96) [inline src/services/API/RootClient.js:68]: why
